src/Stratis.Bitcoin/Consensus/ConsensusManager.cs

```diff
@@ -257,6 +257,12 @@ public ConnectNewHeadersResult HeadersPresented(INetworkPeer peer, List<BlockHea
 
                 connectNewHeadersResult = this.chainedHeaderTree.ConnectNewHeaders(peerId, headers);
 
+                if (!this.peersByPeerId.ContainsKey(peerId))
```

Comment:
The position in the locked code segment that this is placed seems a bit arbitrary (both doing it last and moving it higher up). What was the reasoning behind putting it after `ConnectNewHeaders` in particular instead of, say, just before it?

Reply:
@dangershony didn't wanted to put it before because in case `ConnectNewHeaders` throws the peer will be disconnected, so no need to att to the peersByPeerId dictionary.

Probably it's right because we may incur in a deadlock if we put it before ConnectNewHeaders because it causes peer to disconnect.

Anyway there are more things to investigate about peersByPeerId because it seems this gets populated only for peers supporting PH (after last checkpoint) and this may cause problems

Reply:
@dangershony @MithrilMan I'm also sketchy on why this has been moved up? Care to elaborate?

Reply:
Its a bit long but here goes:

The bellow code is valid scenarios of the flow 
 
From CM.HeadersPresented method
```
                if (connectNewHeadersResult == null)
                {
                    this.logger.LogTrace("(-)[NO_HEADERS_CONNECTED]:null");
                    return null;
                }

                if (connectNewHeadersResult.Consumed == null)
                {
                    this.logger.LogTrace("(-)[NOTHING_CONSUMED]");
                    return connectNewHeadersResult;
                }
```
If both the above conditions are true this means insufficient info to validate a proven header.
This will often happen when the first header is inefficient then  `connectNewHeadersResult.Consumed` will be null (all though @MithrilMan also fixed another bug where we would always report it as null, see the changes to CHT)

If `Consumed == null` causes CMB to do this (basically headers are added to the cache)

```
                if (result.Consumed == null)
                {
                    this.cachedHeaders.AddRange(headers);
                    this.logger.LogDebug("All {0} items were not consumed and added to cache.", headers.Count);

                    return;
                }
```

But this issue is actually reflected later in the method `NotifyBehaviorsOnConsensusTipChangedAsync` this code:
```
     lock (this.peerLock)
            {
                foreach (INetworkPeer peer in this.peersByPeerId.Values)
                    behaviors.Add(peer.Behavior<ConsensusManagerBehavior>());
            }
```

That method will trigger when consensus tip changed and should notify the peer to send the next batch of headers (or actually its proven headers) that are waiting in cache to be processed (because we could not connect them earlier due to missing UTXOs). but the peer was never added to this list `this.peersByPeerId`

So the bug that this change fixes (because the fix to CHT will fix the scenario where any of the headers had insufficient info) is that if we connected to a peer and the first headers it presents could not be connected the peer would be stuck because we would not notify the peer's behaviour to try presenting those headers again when consensus advances.